pkg/ensurance/cm/advanced_cpu_manager.go

```diff
@@ -0,0 +1,351 @@
+
+	// Build a list of (podUID, containerName) pairs for all need to be assigned containers in all active Pods.
+	toBeAssignedContainers := make(map[string]map[string]struct{})
+	for _, pod := range m.activepods() {
+		toBeAssignedContainers[string(pod.UID)] = make(map[string]struct{})
+		for _, container := range pod.Spec.Containers {
+			if m.policy.NeedAllocated(pod, &container) {
+				toBeAssignedContainers[string(pod.UID)][container.Name] = struct{}{}
+				if _, ok := assignments[string(pod.UID)][container.Name]; !ok && doAllocate {
+					err := m.policy.Allocate(m.state, pod, &container)
+					if err != nil {
+						klog.Errorf("[advancedcpumanager] Allocate error: %v", err)
+						continue
+					}
+				}
+			}
+		}
+	}
+
+	// Loop through the CPUManager state. Remove any state for containers not
+	// in the `toBeAssignedContainers` list built above.
+	for podUID := range assignments {
+		for containerName := range assignments[podUID] {
+			if _, ok := toBeAssignedContainers[podUID][containerName]; !ok {
+				klog.Errorf("[advancedcpumanager] removeStaleState: removing (pod %s, container: %s)", podUID, containerName)
+				err := m.policyRemoveContainerByRef(podUID, containerName)
+				if err != nil {
+					klog.Errorf("[advancedcpumanager] removeStaleState: failed to remove (pod %s, container %s), error: %v)", podUID, containerName, err)
+				}
+			}
+		}
+	}
+}
+
+func (m *AdvancedCpuManager) policyRemoveContainerByRef(podUID string, containerName string) error {
+	err := m.policy.RemoveContainer(m.state, podUID, containerName)
+	return err
+}
+
+func (m *AdvancedCpuManager) updateContainerCPUSet(containerID string, cpus cpuset.CPUSet) error {
+	return cruntime.UpdateContainerResources(
+		m.runtimeClient,
+		containerID,
+		cruntime.UpdateOptions{CpusetCpus: cpus.String()},
+	)
+}
+
+func buildContainerMapFromRuntime(runtimeClient pb.RuntimeServiceClient) (containermap.ContainerMap, error) {
+	podSandboxMap := make(map[string]string)
+	podSandboxList, _ := cruntime.ListPodSandboxes(runtimeClient, cruntime.ListOptions{})
+
+	for _, p := range podSandboxList {
+		podSandboxMap[p.Id] = p.Metadata.Uid
+	}
+
+	containerMap := containermap.NewContainerMap()
+	containerList, _ := cruntime.ListContainers(runtimeClient, cruntime.ListOptions{})
+	for _, c := range containerList {
+		if _, exists := podSandboxMap[c.PodSandboxId]; !exists {
+			return nil, fmt.Errorf("no PodsandBox found with Id '%s'", c.PodSandboxId)
+		}
+		containerMap.Add(podSandboxMap[c.PodSandboxId], c.Metadata.Name, c.Id)
+	}
+
+	return containerMap, nil
+}
+
+func GetContainerIdFromPod(pod *v1.Pod, name string) string {
+	if name == "" {
+		return ""
+	}
+
+	for _, v := range pod.Status.ContainerStatuses {
+		if v.Name == name {
+			strList := strings.Split(v.ContainerID, "//")
+			if len(strList) > 0 {
+				return strList[len(strList)-1]
+			}
+		}
+	}
+
+	return ""
+}
+
+func (m *AdvancedCpuManager) AddContainer(p *v1.Pod, c *v1.Container) error {
```

Comment:
Better to sort your function with this order: class public member function -> class private member function -> other functions

Reply:
ok